### PR TITLE
Adding generic preference dataset builder

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

I was writing the docs for the preference datasets but noticed we don't have a generic builder to point to, so, here we are.

#### Changelog
What are the changes made in this PR?
* I've added a `preference_dataset` builder which uses `ChosenToRejectedMessages`. I don't think it's worth adding the stack exchange message format as an option at this time. 
* I've also exposed the HH RLHF dataset builder in our docs.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

File: tests/torchtune/datasets/test_preference_dataset.py [hunk 1]
```diff
@@ -111,3 +112,46 @@ def test_get_item(self, mock_load_dataset, dialogue, expected):
         prompt, label = ds[0]["rejected_input_ids"], ds[0]["rejected_labels"]
         assert prompt == expected_rejected_tokens
         assert label == expected_rejected_labels
+
+    def test_load_local_json(self):
```
Comment (RdoubleA) [inline tests/torchtune/datasets/test_preference_dataset.py:116]: nice, we hoenstly need more of these local dataset tests

File: torchtune/datasets/_preference.py [hunk 2]
```diff
@@ -137,3 +140,136 @@ def _prepare_sample(self, sample: Mapping[str, Any]) -> Dict[str, List[int]]:
...
+                        "role": "user"
+                    },
+                    { "content": "Fix the hole.", "role": "assistant" }
+                ],
+                "rejected": [
+                    {
+                        "content": "What do I do when I have a hole in my trousers?",
+                        "role": "user"
+                    },
+                    { "content": "Take them off.", "role": "assistant" }
```
Comment (RdoubleA) [inline torchtune/datasets/_preference.py:239]: 🫢 🫢 🫢

File: torchtune/datasets/_preference.py [hunk 3]
```diff
@@ -137,3 +140,136 @@ def _prepare_sample(self, sample: Mapping[str, Any]) -> Dict[str, List[int]]:
...
+    >>> from torchtune.datasets import preference_dataset
+    >>> dataset = preference_dataset(
+    ...     tokenizer=tokenizer,
+    ...     source="json",
+    ...     data_files=my_preference_dataset.json,
+    ...     train_on_input=False,
+    ...     split="train",
+    >>> )
+    >>> tokenizer.decode(dataset[0]["chosen_input_ids"], skip_special_tokens=True)
+    What do I do when I have a hole in my trousers?Fix the hole.
```
Comment (RdoubleA) [inline torchtune/datasets/_preference.py:255]: We should fix this trim whitespace issue at some point...

**Review by RdoubleA — APPROVED** [review]
do you want to also enable packing for preference datasets? should be easy to test and there's nothing blocking it

File: torchtune/datasets/_preference.py (hunk 3 above)
Comment (SalmanMohammadi) [reply torchtune/datasets/_preference.py:260]: actually renders for me as `'user\n\nWhat do I do when I have a hole in my trousers?assistant\n\nTake them off.'` with llama3 tokenizer and `skip_special_tokens=True`, and `'[INST] What do I do when I have a hole in my trousers?  [/INST] Take them off.'` with llama2, but I wrote it this way for consistency with the other examples.

**Comment by SalmanMohammadi** [comment]
> do you want to also enable packing for preference datasets? should be easy to test and there's nothing blocking it

I'm actually pretty out of the loop on packed datasets, sorry. TLDR how to?

**Comment by RdoubleA** [comment]
> I'm actually pretty out of the loop on packed datasets, sorry. TLDR how to?

just emulate this and that's all you need, then expose the packed parameter: https://github.com/pytorch/torchtune/blob/c5db813ce0473db090a4f1f6b450f559acac58e5/torchtune/datasets/_instruct.py#L257

then just run a DPO config with preference_dataset and use the override `dataset.packed=True tokenizer.max_seq_len=4096` and see if it packs successfully and trains. You can hardcode num packs to make it pack faster
